src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller.cs

```diff
@@ -224,37 +222,53 @@ private OptionSet GetOptions()
                 : workspace.Options;
         }
 
-        private void CommitItem(CompletionItem item)
+        private void CommitItem(PresentationItem item)
         {
             AssertIsForeground();
 
-            item = Controller.GetExternallyUsableCompletionItem(item);
-
             // We should not be getting called if we didn't even have a computation running.
             Contract.ThrowIfNull(this.sessionOpt);
             Contract.ThrowIfNull(this.sessionOpt.Computation.InitialUnfilteredModel);
 
             // If the selected item is the builder, there's not actually any work to do to commit
-            if (item.IsBuilder)
+            if (item.IsSuggestionModeItem)
             {
                 this.StopModelComputation();
                 return;
             }
 
-            var textChange = GetCompletionRules().GetTextChange(item);
-            this.Commit(item, textChange, this.sessionOpt.Computation.InitialUnfilteredModel, null);
+            this.Commit(item, this.sessionOpt.Computation.InitialUnfilteredModel, commitChar: null);
+        }
+
+        private const int MaxMRUSize = 10;
+        private ImmutableArray<string> _recentItems = ImmutableArray<string>.Empty;
+
+        public void MakeMostRecentItem(string item)
+        {
+            bool updated = false;
+
+            while (!updated)
```

Comment:
Why the "while + InterlockedUpdate" approach?  THis seems like a simple lock would be sufficient for implementing this...